Coalesced RDD with locality

alig opened the pull request:
Coalesced RDD which tries to pack the parents so that there exists one preferred location for the coalesced RDD that most of the coalesced partitions have in common.

alig committed:
load balancing coalescer

alig committed:
Bug, should compute slack wrt parent partition size, not number of bins

alig committed:
Large scale load and locality tests for the coalesced partitions added

alig committed:
added goals inline as comment

AmplabJenkins commented:
Thank you for your pull request. An admin will review this request soon.

alig committed:
word wrap before 100 chars per line

rxin commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -19,11 +19,13 @@ package spark.rdd
     @transient rdd: RDD[_],
-    parentsIndices: Array[Int]
+    parentsIndices: Array[Int],
+    prefLoc: String = ""
```
Make this a transient val, and rename it preferredLoc. Then you can remove the getPreferredLocation method.

alig replied:
Done

rxin commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -33,6 +35,24 @@ private[spark] case class CoalescedRDDPartition(
     parents = parentsIndices.map(rdd.partitions(_))
```
This is not your change. But I am wondering if this parents = ... is needed because in the constructor, we are already setting the parents.

alig replied:
Agree. I'm leaving it for now because I don't want to break anything. But since it's an immutable, setting it once in the constructor should be sufficient.

rxin commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -33,6 +35,24 @@ private[spark] case class CoalescedRDDPartition(
+   */
+  def localFraction :Double = {
+    var loc: Int = 0
+    parents.foreach(p => if (rdd.preferredLocations(p).contains(getPreferredLocation)) loc += 1)
```
Nice handling of error cases. The first two lines we can simplify it a little bit to

```
val loc = parents.count(p => rdd.preferredLocations(p).contains(getPreferredLocation))
```

alig replied:
Done

rxin commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -42,23 +62,44 @@ private[spark] case class CoalescedRDDPartition(
+ * If there is locality information, it proceeds to pack them with the following three goals:
+ *
+ * (1) Balance the groups so they roughly have the same number of parent partitions
+ * (2) Achieve locality per partition, i.e. there exists one machine which most parent splits prefer
```
Let's change all references to "split" to "partition". "split" was a term we used previously. Now we are using the term "partition" only. In source code, you still see the use of "split" because they are old code, but we definitely should avoid using "split" in documentation.

alig replied:
Done

mateiz commented:
Jenkins, add to whitelist and test this please

rxin commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -42,23 +62,44 @@ private[spark] case class CoalescedRDDPartition(
     @transient var prev: RDD[T],
-    maxPartitions: Int)
+    maxPartitions: Int,
+    balanceSlack: Double = 0.10 )
```
Remove the space before )

alig replied:
done

rxin commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -42,23 +62,44 @@ private[spark] case class CoalescedRDDPartition(
+    val res = mutable.ArrayBuffer[CoalescedRDDPartition]()
+    val packer = new PartitionCoalescer(maxPartitions, prev, balanceSlack)
+
+    for ((pg, i) <- packer.getPartitions.zipWithIndex) {
```
Instead of a for loop, if you use a map, you can get rid of the ArrayBuffer

```
packer.getPartitions.zipWithIndex.map { case(pg, i) =>
  ...
}.toArray
```

alig replied:
Good catch, done!

rxin commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +119,209 @@ class CoalescedRDD[T: ClassManifest](
+  override def getPreferredLocations(split: Partition): Seq[String] = {
+    if (split.isInstanceOf[CoalescedRDDPartition])
+      List(split.asInstanceOf[CoalescedRDDPartition].getPreferredLocation)
+    else
```
Why calling super? We should just return Nil if the partition is not of proper type. Also, for if/else statements, always wrap them with curly braces.

alig replied:
Thanks, wrapped the if/else's in `{ }`.

We are overriding the super class. The type checker will not catch giving it just an actual Partition, so I'm just deferring that to the base class which can implement it however it wants. Say it wants to throw an exception or something in the future.

rxin commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +119,209 @@ class CoalescedRDD[T: ClassManifest](
+        for ((p,i) <- prev.partitions.zipWithIndex) {
+          groupArr(i).list += p
+        }
+      } else { // old code, just splits them grouping partitions that are next to each other
```
remove "old old" since most people won't have a clue what old cold means here.

alig replied:
Done

rxin commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +119,209 @@ class CoalescedRDD[T: ClassManifest](
+
+    // initializes/resets to start iterating from the beginning
+    private def resetIterator() = {
+      val i1 =  prev.partitions.view.map( (p: Partition) =>
```
Write map's like this

```
val i1 =  prev.partitions.view.map { p: Partition =>
  ....
}
```

alig replied:
Thanks, will do (and done). That also allows me to use uniform syntax for partial functions, indeed useful!

rxin commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +119,209 @@ class CoalescedRDD[T: ClassManifest](
+      pgroup.list += part           // already assign this element
+      initialHash += (part -> true) // needed to avoid assigning partitions to multiple buckets
+      true
+    } else false
```
wrap the else with { }. ditto for other if/else statements too

alig replied:
Done here and elsewhere, except in one liners, say:

``` scala
val ratio = if (denominator==0) 0 else numerator/denominator
```

Can change those too if you prefer, but it is concise and unlikely that someone by mistakes makes it multi-line without braces.

rxin commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +119,209 @@ class CoalescedRDD[T: ClassManifest](
+}
+
+
+class PartitionCoalescer(maxPartitions: Int, prev: RDD[_], balanceSlack: Double) {
```
In general, this looks like a single funciton. Maybe we should restructure it to turn this into an Object, and define a coalesce function the CoalescedRDD can call?

alig replied:
Good idea, created a functional object.

rxin commented:
Hi Ali - this looks pretty good. Thanks for doing it. I left some comments on style and design. I didn't look closely into the algorithm itself though.

AmplabJenkins commented:
Thank you for submitting this pull request. 

Unfortunately, the automated tests for this request have failed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/588/

alig commented:
Thanks for taking time and commenting on this. I'll fix them ASAP.

--Ali

On Aug 14, 2013, at 9:37 PM, Reynold Xin notifications@github.com wrote:

> Hi Ali - this looks pretty good. Thanks for doing it. I left some comments on style and design. I didn't look closely into the algorithm itself though.
> 
> —
> Reply to this email directly or view it on GitHub.

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/591/

alig committed:
Renamed split to partition

alig committed:
Fix bug in tests

alig committed:
made preferredLocation a val of the surrounding case class

alig committed:
use count rather than foreach

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/592/

alig committed:
space removed

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/593/

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/594/

alig committed:
several of Reynold's suggestions implemented

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/596/

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/597/

alig committed:
Made a function object that returns the coalesced groups

mridulm commented:
It would be better to abstract out the functionality in this PR and reuse it elsewhere too.
For example, shuffle can leverage rack locality of map output locations to optimize on minimal cross rack local traffic for example.

AmplabJenkins commented:
Thank you for submitting this pull request. 

Unfortunately, the automated tests for this request have failed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/601/

rxin commented:
That's a good idea. Let's do it in a separate patch later.

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -19,20 +19,32 @@ package spark.rdd
+   * as one of their preferredLocations
+   * @return locality of this coalesced partition between 0 and 1
+   */
+  def localFraction :Double = {
```
Formatting is weird here.. should be `localFraction: Double`

alig replied:
Yeah weird, fixed! Thanks.

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +109,215 @@ class CoalescedRDD[T: ClassManifest](
+  def size = list.size
+}
+
+object coalescePartitions {
```
This is not a good name format for an object or class -- it should start with a capital letter, and be a noun. Call it PartitionCoalescer maybe. Also, I don't think we need an `object` for it.. users could call it with just `new PartitionCoalescer().run(stuff)`.

mateiz replied:
Also, mark this and the class above as `private[spark]`.

alig replied:
Done. Changed name back to the original PartitionCoalescer and added the functionality in run() and marked classes as private[spark]

mateiz commented:
Hey, one comment on this -- it's currently using the parent RDD's `preferredLocations` to determine where to run, which will only work if you run `coalesce` directly on a `HadoopRDD`, without any maps or other operations in-between. This is because preferred locations are not inherited down the chain. To make this work in more general cases, where I think it will still be quite useful, it should use `DAGScheduler.getPreferredLocs`. @rxin, is that okay? What is the use case in Shark (what is the lineage graph you call coalesce on)?

rxin commented:
The current use case is to operate directly on a hadoop rdd. I agree in general it'd be better to call getPreferredLocs from DAGScheduler (actually we need to create a recursive getPreferredLocs similar to the one in DAGScheduerl, but shouldn't take cache locality into account). 

This one as is is already a big improvement. We can perhaps change to getPreferredLocs in a later PR.

alig committed:
fixed matei's comments

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/624/

alig committed:
Added error messages to the tests to make failed tests less cryptic

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/626/

alig committed:
added one test that will test a future functionality

alig committed:
Coalescer now uses current preferred locations for derived RDDs. Made run() in DAGScheduler thread safe and added a method to be able to ask it for preferred locations. Added a similar method that wraps the former inside SparkContext.

alig committed:
comment in the test to make it more understandable

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/629/

alig commented:
Ok, implemented functionality to augment `SparkContext` and `DAGScheduler` with methods to get the _current_ preferred locations of derived RDDs. I made the extensions to `DAGScheduler` thread safe. Also added tests that now test whether derived RDDs are scheduled for locality too. Please have a look @mateiz and @rxin.

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/636/

alig committed:
making CoalescedRDDPartition public

rxin commented on core/src/main/scala/spark/scheduler/DAGScheduler.scala:
```diff
@@ -774,7 +777,13 @@ class DAGScheduler(
+   * @param partition to lookup locality information for
+   * @return list of machines that are preferred by the partition
+   */
+  private[spark] def getPreferredLocs(rdd: RDD[_], partition: Int): List[String] = synchronized {
```
I think it is best to just implement a recursive version of getPreferredLocs in the RDD class itself, without taking memory locality into account there. There are two reasons:
1. This method is called pretty often in the scheduler, and can become a pretty serious bottleneck if we synchronized it.
2. I don't think we should use memory locality in the coalesced rdd's preferred locations, because the life span of a coalesced rdd's preferred locations is the same as the rdd itself, but the life span of the cache locations vary.

mateiz replied:
I actually disagree with this -- although cache locations might change, we should make sure we have sensible behavior when people do want to coalesce a cached RDD. For example, a common use case I can imagine is doing a filter or local reduce on a big cached RDD, then coalescing the result of that to be more efficient to deal with. In most cases (e.g. HDFS files) we'll be coalescing non-cached RDDs anyway and only caching after the coalesce.

The synchronized here shouldn't be a problem. The DAGScheduler event loop is single-threaded so it will have no trouble acquiring the lock, not to mention it only acquires the lock once per processEvent() call, not separately on each getPreferredLocs() call.

mateiz replied:
BTW, one other question on this -- have you tested it with 100K partitions again? If it's slow, then one solution would be to add a "bulk" version of getPreferredLocs that returns the locations for all partitions of an RDD at once as an array of lists, and do this just once in PartitionCoalescer, then look into that array instead of calling this.

alig replied:
Just for the record, 100k partitions coalesced into 2k takes 850ms on my MBA. Doing it in bulk didn't buy much, perhaps 10ms on average (repeating the experiment 20 times to get the std.dev down to negligible numbers). So for now, we're making one call to DAGScheduler for each partition of the RDD.

mateiz commented on core/src/main/scala/spark/SparkContext.scala:
```diff
@@ -616,6 +616,15 @@ class SparkContext(
+   * @param partition to be looked up for locality
+   * @return list of preferred locations for the partition
+   */
+  def getPreferredLocs(rdd: RDD[_], partition: Int): List[String] =
```
This should be `private[spark]`

alig replied:
Done

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -17,22 +17,37 @@
+   */
+  def localFraction: Double = {
+    val loc = parents.count(p => rdd.context.getPreferredLocs(rdd, p.index)
+      .contains(preferredLocation))
```
Would look better as

```
val loc = parents.count(p =>
  rdd.context.getPreferredLocs(rdd, p.index).contains(preferredLocation))
```

alig replied:
Done

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+    } else {
+      super.getPreferredLocations(partition)
+    }
+  }
```
Why would the partition passed not be a CoalescedRDDPartition? Have you seen this happen?

alig replied:
Nope, just defensive programming since the compiler wouldn't be able to catch that. Let me know if I should just assume it is.

mateiz replied:
Yeah you should, because I think passing something than the RDD's own Partition objects here would be a logic failure somewhere else, and we should catch it.

alig replied:
Ok, so should I just assume it's the right type or should I check for it and throw an exception? Note that the parent anyway just returns Nil, so I could just return Nil too. Just let me know.

mateiz replied:
Yes, just cast it as if it were the right type.

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -42,28 +57,47 @@ private[spark] case class CoalescedRDDPartition(
+ * (2n log(n)). The load balancing is done using power-of-two randomized bins-balls with one twist:
+ * it tries to also achieve locality. This is done by allowing a slack (balanceSlack) between two
+ * bins. If two bins are within the slack in terms of balance, the algorithm will assign partitions
+ * according to locality. (contact alig for questions)
```
This is a great doc comment on the algorithm but I'd move it to the `PartitionCoalescer` class to put it closer to the code. In `CoalescedRDD` you can just say it tries to take into account both locality and load-balancing.

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+}
+
+private[spark] class PartitionCoalescer(maxPartitions: Int, prev: RDD[_],
+                                        balanceSlack: Double) {
```
This can fit on a 100-character line. If it doesn't, move the `private[spark]` to a line by itself and leave the `class` stuff below.

alig replied:
Done

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+
+private[spark] class PartitionCoalescer(maxPartitions: Int, prev: RDD[_],
+                                        balanceSlack: Double) {
+  protected def compare(o1: PartitionGroup, o2: PartitionGroup): Boolean = o1.size < o2.size
```
The members of this class probably don't need to be protected, because we won't have people extending it. You can just remove those protected annotations to reduce clutter.

alig replied:
Done

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+  protected val rnd = new scala.util.Random(7919) // keep this class deterministic
+
+  // each element of groupArr represents one coalesced partition
+  protected val groupArr = mutable.ArrayBuffer[PartitionGroup]()
```
Just import ArrayBuffer and ListBuffer instead of prefixing them with mutable. IntelliJ recommends using the prefix for Map because there's also an immutable.Map, so that's fine, but these guys have no immutable counterparts.

mateiz replied:
Also, BTW, there's probably no reason to use ListBuffer over ArrayBuffer in this code

alig replied:
Done!

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+  // next() returns the next preferred machine that a partition is replicated on
+  // the rotator first goes through the first replica copy of each partition, then second, third
+  // the iterators return type is a tuple: (replicaString, partition)
+  protected class RotateLocations(prev: RDD[_]) extends Iterator[(String, Partition)] {
```
Name this LocationIterator to make it clearer that it's an interator

alig replied:
Done

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+      { if (currentPreferredLocations(prev, p).length > 2)
+        Some((currentPreferredLocations(prev, p)(2),p)) else None } }
+      val res = List(i1,i2,i3)
+      res.view.flatMap(x => x).flatten.iterator // fuses the 3 iterators (1st replica, 2nd, 3rd)
```
If you make i1, i2 and such iterators, you can just do `i1 ++ i2 ++ i3`

alig replied:
Done

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+    protected def resetIterator() = {
+      val i1 =  prev.partitions.view.map{ p: Partition =>
+      { if (currentPreferredLocations(prev, p).length > 0)
+        Some((currentPreferredLocations(prev, p)(0),p)) else None } }
```
Just make these iterators instead of views, e.g.:

```
val i1 = prev.partitions.iterator.map(p => currentPreferredLocations(prev, p)).filter(_.length > 0).map(s => s(0))
```

Also since `currentPreferredLocations` is only called on `prev` as the RDD, I'd make it just use that by default instead of having a parameter.

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+
+    // return the next preferredLocation of some partition of the RDD
+    def next(): (String, Partition) = {
+      if (it.hasNext)
```
Put braces around the if body even if it's one line

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+  protected val groupHash = mutable.Map[String, mutable.ListBuffer[PartitionGroup]]()
+
+  // hash used for the first maxPartitions (to avoid duplicates)
+  protected val initialHash = mutable.Map[Partition, Boolean]()
```
If you just use it to test membership, this can be a Set instead of Map

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+    }
+
+    while (numCreated < targetLen) {  // if we don't have enough partition groups, create duplicates
+    var (nxt_replica, nxt_part) = rotIt.next()
```
Indent seems wrong

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+          groupArr(i).list += p
+        }
+      } else { // no locality available, then simply split partitions based on positions in array
+        (0 until maxPartitions).foreach { i =>
```
Clearer to write this as `for (i <- 0 until maxPartitions)`; same with the foreach over a range below

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+}
+
+private[spark] case class PartitionGroup(prefLoc: String = "") {
+  var list = mutable.ListBuffer[Partition]()
```
Same here, probably no need for ListBuffer instead of ArrayBuffer.

mateiz commented on core/src/main/scala/spark/rdd/CoalescedRDD.scala:
```diff
@@ -78,4 +112,221 @@ class CoalescedRDD[T: ClassManifest](
+    }
+  }
+
+  protected def getPartitions : Array[PartitionGroup] = groupArr.filter( pg => pg.size > 0).toArray
```
Don't put a space before colons

mateiz commented on core/src/test/scala/spark/RDDSuite.scala:
```diff
@@ -173,6 +174,61 @@ class RDDSuite extends FunSuite with SharedSparkContext {
+    assert(coalesced1.collect().toList.sorted === (1 to 9).toList, "Data got *lost* in coalescing")
+
+    val splits = coalesced1.glom().collect().map(_.toList).toList
+    assert(splits.length == 3, "Supposed to coalesce to 3 but got " + splits.length)
```
Use `===` instead of `==`

mateiz commented on core/src/test/scala/spark/RDDSuite.scala:
```diff
@@ -173,6 +174,61 @@ class RDDSuite extends FunSuite with SharedSparkContext {
+    assert(splits.length == 3, "Supposed to coalesce to 3 but got " + splits.length)
+
+    assert(splits.foldLeft(true)
+      ( (x,y) => if (!x) false else y.length >= 1) === true, "Some partitions were empty")
```
Simpler to write this as `splits.forall(_.length >= 1)`

mateiz commented on core/src/test/scala/spark/RDDSuite.scala:
```diff
@@ -173,6 +174,61 @@ class RDDSuite extends FunSuite with SharedSparkContext {
+    val coalesced4 = data.coalesce(20)
+    assert(coalesced4.glom().collect().map(_.toList).toList.sortWith(
+      (x, y) => if (x.isEmpty) false else if (y.isEmpty) true else x(0) < y(0)) === (1 to 9).
+      map(x => List(x)).toList, "Tried coalescing 9 partitions to 20 but didn't get 9 back")
```
This is a super-long expression.. split it into smaller ones.

mateiz commented on core/src/test/scala/spark/RDDSuite.scala:
```diff
@@ -173,6 +174,61 @@ class RDDSuite extends FunSuite with SharedSparkContext {
+    (1 to numMachines).foreach(machines += "m"+_)
+
+    val blocks = (1 to partitions).map( i => (i, (i to (i+2))
+      .map{ j => machines(rnd.nextInt(machines.size)) } ))
```
The line splitting here is not great; it makes it hard to see that the .map is on the range i to i+2, and all this stuff is in a tuple. Feel free to either use temporary variables or use blocks, e.g.

```
(1 to partitions).map { i =>
  val locs = (i to (i+2)).map(_ => machines(rnd.nextInt(machines.size))
  (i, locs)
}
```

mateiz replied:
Also, if this is just using 3 random hosts, there's no point using (i to (i+2)).. you can use (1 to 3) or just `Array.fill(3)(machines(rnd.nextInt(machines.size))`

mateiz commented on core/src/test/scala/spark/RDDSuite.scala:
```diff
@@ -173,6 +174,61 @@ class RDDSuite extends FunSuite with SharedSparkContext {
+
+    // test that you get over 90% locality in each group
+    val minLocality = coalesced2.partitions
+      .map( part => part.asInstanceOf[CoalescedRDDPartition].localFraction )
```
Don't leave spaces after open parens or before close parens

mateiz commented on core/src/test/scala/spark/RDDSuite.scala:
```diff
@@ -173,6 +174,61 @@ class RDDSuite extends FunSuite with SharedSparkContext {
+      map(x => List(x)).toList, "Tried coalescing 9 partitions to 20 but didn't get 9 back")
+
+
+    // large scale experiment
```
Instead of putting this at the bottom of a big "cogrouped RDD with locality" test, make it a separate test case. It will make it easier to see what's failing.

mateiz commented on core/src/test/scala/spark/RDDSuite.scala:
```diff
@@ -173,6 +174,61 @@ class RDDSuite extends FunSuite with SharedSparkContext {
+      .foldLeft(0)((dev, curr) => math.max(math.abs(100-curr),dev))
+    assert(maxImbalance <= 20, "Expected 100 +/- 20 per partition, but got " + maxImbalance)
+
+    // TDD: Test for later when we have implemented functionality to get locality from DAGScheduler
```
This is done now, so no need to keep this comment, right?

mateiz commented:
Hey Ali, I've looked through this now too and made some detailed comments, almost entirely on Scala style. The actual algorithm and tests look good. Thanks for putting this together!

alig committed:
fixed Matei's comments

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/649/

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/652/

alig committed:
Fixed almost all of Matei's feedback

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/657/

alig committed:
simpler code

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/660/

alig committed:
Simpler code

alig committed:
Added a test to make sure no locality preferences are ignored

alig committed:
Bug in test fixed

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/667/

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/668/

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/669/

alig committed:
indent

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/677/

alig commented:
Done with all of Matei's comments.

alig committed:
added curly braces to make the code more consistent

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/678/

alig committed:
Merged in from upstream to use TaskLocation instead of strings

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/679/

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/680/

AmplabJenkins commented:
Thank you for submitting this pull request. 

Unfortunately, the automated tests for this request have failed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/681/

mateiz commented:
Jenkins, please retest this

mateiz commented:
Jenkins, retest this please

mateiz commented:
@alig this looks good to me, the only thing I'd like to do is run the scheduling throughput perf-test to make sure it doesn't get affected. After that I'll merge it.

alig commented:
Sure @mateiz, I'll do that report it here.

AmplabJenkins commented:
Thank you for submitting this pull request. 

Unfortunately, the automated tests for this request have failed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/685/

alig commented:
Jenkins, retest this please

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/687/

alig commented:
Ok, so ran the performance tests for the master branch and this pull-request:

Results for the current master branch:

``` bash
scheduling-tput:      5.603,  6.545,  4.659,  4.659, 21.521
scala-agg-by-key:    29.658, 17.575, 21.255, 21.255, 69.928
scala-sort-by-key:    9.269,  0.765,  9.214,  9.214,  11.19
scala-count:         17.582, 17.169,  6.689,  6.689, 55.479
scala-count-w-fltr:  15.128, 17.194,  7.087,  7.087, 55.917
```

Results for this pull-request:

``` bash
scheduling-tput:      4.703,  4.719,  4.522,  4.522, 16.616
scala-agg-by-key:    30.365, 20.058,  21.46,  21.46, 75.604
scala-sort-by-key:    9.184,  0.719,  8.932,  8.932, 10.904
scala-count:         17.852, 20.271,  6.349,  6.349, 54.761
scala-count-w-fltr:  14.886, 16.503, 13.971, 13.971, 56.561
```

The numbers are in order: median, std.dev, min, first, last. 
As a side note, I did actually take time to check carefully that the two different performance tests were using their respective branches of the source code and not just using the same source twice.

mateiz commented:
OK, that looks like it didn't break anything. Going to merge it then.
